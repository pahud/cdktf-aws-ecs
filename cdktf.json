{
  "language": "typescript",
  "app": "npx ts-node src/integ.default.ts",
  "terraformModules": [
    "terraform-aws-modules/vpc/aws",
    "terraform-aws-modules/ecs/aws"
  ],
  "codeMakerOutput": "src/imports",
  "projectId": "9106836c-8643-4ffb-a965-96d40c3436f2"
}