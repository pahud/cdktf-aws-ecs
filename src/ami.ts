import { Construct } from 'constructs';
import { StringParameter } from '.';

export enum AmazonLinuxGeneration {
  AMAZON_LINUX,
  AMAZON_LINUX_2,
}

export enum AmiHardwareType {
  /**
   * Use the standard Amazon ECS-optimized AMI.
   */
  STANDARD = 'Standard',

  /**
   * Use the Amazon ECS GPU-optimized AMI.
   */
  GPU = 'GPU',

  /**
   * Use the Amazon ECS-optimized Amazon Linux 2 (arm64) AMI.
   */
  ARM = 'ARM64',
}

/**
 * Identifies an instance's CPU architecture
 */
export enum InstanceArchitecture {
  /**
   * ARM64 architecture
   */
  ARM_64 = 'arm64',

  /**
   * x86-64 architecture
   */
  X86_64 = 'x86_64',
}

export interface IEcsMachineImage {
  get amiId(): string;
}

export interface EcsOptimizedAmiProps {
  readonly generation?: AmazonLinuxGeneration;
  readonly hwtype?: AmiHardwareType;
}

export class EcsOptimizedAmi implements IEcsMachineImage {
  readonly amiParameterName: string;
  readonly generation: AmazonLinuxGeneration;
  readonly hwType: AmiHardwareType;
  readonly scope: Construct;
  constructor(scope: Construct, props?: EcsOptimizedAmiProps) {
    this.generation = props?.generation ?? AmazonLinuxGeneration.AMAZON_LINUX_2,
    this.hwType = props?.hwtype ?? AmiHardwareType.STANDARD,
    this.scope = scope;
    // set the SSM parameter name
    this.amiParameterName = '/aws/service/ecs/optimized-ami/'
      + (this.generation === AmazonLinuxGeneration.AMAZON_LINUX ? 'amazon-linux/' : '')
      + (this.generation === AmazonLinuxGeneration.AMAZON_LINUX_2 ? 'amazon-linux-2/' : '')
      + (this.hwType === AmiHardwareType.GPU ? 'gpu/' : '')
      + (this.hwType === AmiHardwareType.ARM ? 'arm64/' : '')
      + 'recommended/image_id';
  }
  /**
   * Return the correct image
   */
  get amiId(): string {
    return lookupImage(this.scope, 'EcsOptimizedAmiId', this.amiParameterName);
  }
}

/**
 * Amazon ECS variant
 */
export enum BottlerocketEcsVariant {
  /**
   * aws-ecs-1 variant
   */
  AWS_ECS_1 = 'aws-ecs-1'

}

/**
 * Properties for BottleRocketImage
 */
export interface BottleRocketImageProps {
  /**
   * The Amazon ECS variant to use.
   * Only `aws-ecs-1` is currently available
   *
   * @default - BottlerocketEcsVariant.AWS_ECS_1
   */
  readonly variant?: BottlerocketEcsVariant;

  /**
   * The CPU architecture
   *
   * @default - x86_64
   */
  readonly architecture?: InstanceArchitecture;

  /**
   * Whether the AMI ID is cached to be stable between deployments
   *
   * By default, the newest image is used on each deployment. This will cause
   * instances to be replaced whenever a new version is released, and may cause
   * downtime if there aren't enough running instances in the AutoScalingGroup
   * to reschedule the tasks on.
   *
   * If set to true, the AMI ID will be cached in `cdk.context.json` and the
   * same value will be used on future runs. Your instances will not be replaced
   * but your AMI version will grow old over time. To refresh the AMI lookup,
   * you will have to evict the value from the cache using the `cdk context`
   * command. See https://docs.aws.amazon.com/cdk/latest/guide/context.html for
   * more information.
   *
   * Can not be set to `true` in environment-agnostic stacks.
   *
   * @default false
   */
  readonly cachedInContext?: boolean;
}

/**
 * Construct an Bottlerocket image from the latest AMI published in SSM
 */
export class BottleRocketImage implements IEcsMachineImage {
  private readonly amiParameterName: string;
  /**
   * Amazon ECS variant for Bottlerocket AMI
   */
  private readonly variant: string;

  /**
   * Instance architecture
   */
  private readonly architecture: InstanceArchitecture;

  private readonly scope: Construct;

  // private readonly cachedInContext: boolean;

  /**
   * Constructs a new instance of the BottleRocketImage class.
   */
  public constructor(scope: Construct, props: BottleRocketImageProps = {}) {
    this.scope = scope;
    this.variant = props.variant ?? BottlerocketEcsVariant.AWS_ECS_1;
    this.architecture = props.architecture ?? InstanceArchitecture.X86_64;

    // set the SSM parameter name
    this.amiParameterName = `/aws/service/bottlerocket/${this.variant}/${this.architecture}/latest/image_id`;

    // this.cachedInContext = props.cachedInContext ?? false;
  }

  /**
   * Return the correct image
   */
  get amiId(): string {
    return lookupImage(this.scope, 'BottlerocketAmiId', this.amiParameterName);
  }

  /**
   * Return the correct image
   */
  // public getImage(scope: CoreConstruct): ec2.MachineImageConfig {
  // const ami = lookupImage(scope, this.cachedInContext, this.amiParameterName);

  // return {
  //   imageId: ami,
  //   osType: ec2.OperatingSystemType.LINUX,
  //   userData: ec2.UserData.custom(''),
  // };
  // }
}


function lookupImage(scope: Construct, id: string, parameterName: string) {
  return StringParameter.valueFromLookup(scope, id, parameterName);
}
