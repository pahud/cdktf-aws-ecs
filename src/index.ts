export * from './cluster';
export * from './ssm';
export * from './ami';
