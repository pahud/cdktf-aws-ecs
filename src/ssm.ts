import { DataAwsSsmParameter } from '@cdktf/provider-aws';
import { Construct } from 'constructs';

export class StringParameter {
  static valueFromLookup(scope: Construct, id: string, parameterName: string): string {
    return new DataAwsSsmParameter(scope, id, {
      name: parameterName,
    }).value;
  }
}
