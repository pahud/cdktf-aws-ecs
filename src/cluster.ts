import { AwsProvider, AutoscalingGroup, EcsCapacityProvider, LaunchTemplate, IamRole, IamInstanceProfile, IamPolicyAttachment, DataAwsAvailabilityZones as AZ } from '@cdktf/provider-aws';
import { Token } from 'cdktf';
import { Construct } from 'constructs';
import { AmazonLinuxGeneration, EcsOptimizedAmi } from '.';
import * as awsEcs from './imports/modules/terraform-aws-modules/ecs/aws';
import * as awsVpc from './imports/modules/terraform-aws-modules/vpc/aws';


export class ClusterProps {
  /**
   * The AWS region to deploy.
   */
  readonly region?: string;
  readonly name?: string;
  /**
   * instance type for the default capacity
   * @default t3.large
   */
  readonly instanceType?: string;
  /**
   * VPC subnet IDs for the container instances
   * @default - private subnets of a new VPC.
   */
  readonly vpcSubnets?: string[];
}

export interface AsgCapacityOptions {
  readonly maxCapacity?: number;
  readonly minCapacity?: number;
  readonly desiredCapacity?: number;
}

export class Cluster extends Construct {
  readonly region: string;
  readonly clusterName: string;
  readonly vpc?: any;
  readonly vpcId?: string;
  private readonly id: string;
  private readonly vpcSubnets: string[];
  private readonly props: ClusterProps;
  constructor(scope: Construct, id: string, props: ClusterProps = {}) {
    super(scope, id);

    this.id = id;
    this.props = props;
    this.region = props.region ?? 'us-east-1';
    new AwsProvider(this, 'aws', { region: this.region });

    this.clusterName = props.name ?? `${id}-cluster`;

    // no private subnets given
    if (!props.vpcSubnets) {
      const vpc = this._createVpc();
      this.vpc = vpc;
      this.vpcId = Token.asString(vpc.vpcIdOutput);
      this.vpcSubnets = Token.asList(vpc.privateSubnetsOutput);
    } else {
      this.vpcSubnets = props.vpcSubnets!;
    }

    // cluster
    new awsEcs.TerraformAwsModulesEcsAws(this, 'Cluster', {
      name: this.clusterName,
    });
  }
  private _createContainerServiceforEC2Role(): IamRole {
    const role = new IamRole(this, 'ecs-instance-role', {
      name: `${this.id}-ecs-instance-role`,
      // name: 'ecs-instance-role',
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Sid: '',
            Principal: {
              Service: 'ec2.amazonaws.com',
            },
          },
        ],
      }),
    });
    new IamPolicyAttachment(this, ' AmazonEC2ContainerServiceforEC2RoleAttachment', {
      name: 'AmazonEC2ContainerServiceforEC2RoleAttachment',
      policyArn: 'arn:aws:iam::aws:policy/service-role/AmazonEC2ContainerServiceforEC2Role',
      roles: [role.name],
    });
    return role;
  }
  private _createIamInstanceProfile(roleName: string): IamInstanceProfile {
    return new IamInstanceProfile(this, 'InstanceProfile', {
      role: roleName,
    });
  }
  private _createVpc() {
    const vpc = new awsVpc.TerraformAwsModulesVpcAws(this, 'Vpc', {
      cidr: '10.0.0.0/16',
      azs: new AZ(this, 'AZs', {
        state: 'available',
      }).names,
      publicSubnets: ['10.0.1.0/24', '10.0.2.0/24', '10.0.3.0/24'],
      privateSubnets: ['10.0.11.0/24', '10.0.12.0/24', '10.0.13.0/24'],
      singleNatGateway: true,
      enableNatGateway: true,
      oneNatGatewayPerAz: false,
    });
    return vpc;
  }
  private _createLaunchTemplate(instanceProfileArn: string): LaunchTemplate {
    return new LaunchTemplate(this, 'LT', {
      imageId: new EcsOptimizedAmi(this, {
        generation: AmazonLinuxGeneration.AMAZON_LINUX_2,
      }).amiId,
      iamInstanceProfile: [{ arn: instanceProfileArn }],
      instanceType: this.props.instanceType ?? 't3.large',
    });
  }
  public addAsgCapacity(id: string, options: AsgCapacityOptions) {
    const instanceRole = this._createContainerServiceforEC2Role();
    const instanceProfile = this._createIamInstanceProfile(instanceRole.name);
    const lt = this._createLaunchTemplate(instanceProfile.arn);
    let userData = `#!/bin/bash\necho ECS_CLUSTER=${this.clusterName} > /etc/ecs/ecs.config`;
    let b64userdata = Buffer.from(userData).toString('base64');
    lt.addOverride('user_data', b64userdata);

    const minCapacity = options.minCapacity ?? 0;
    const desiredCapacity = options.desiredCapacity ?? minCapacity;
    const maxCapacity = options.maxCapacity ?? (desiredCapacity > 0) ?
      desiredCapacity : 1;

    const asg = new AutoscalingGroup(this, `${id}-ASG`, {
      maxSize: maxCapacity,
      desiredCapacity: desiredCapacity,
      minSize: minCapacity,
      vpcZoneIdentifier: this.vpcSubnets,
      launchTemplate: [
        {
          id: lt.id,
        },
      ],
    });
    new EcsCapacityProvider(this, 'EcsCapacityProvider', {
      name: `cp-${this.id}`,
      autoScalingGroupProvider: [
        {
          autoScalingGroupArn: asg.arn,
        },
      ],
    });
  }
}
