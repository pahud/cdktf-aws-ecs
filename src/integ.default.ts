import { App, TerraformStack } from 'cdktf';
import { BottleRocketImage, Cluster } from '.';

const app = new App();

const stack = new TerraformStack(app, 'cdktf-eks-demo');

const cluster = new Cluster(stack, 'EcsCluster');

cluster.addAsgCapacity('DefaultCapacity', {
  maxCapacity: 10,
  minCapacity: 0,
  desiredCapacity: 2,
});

cluster.addAsgCapacity('BRCapacity', {
  maxCapacity: 10,
  minCapacity: 0,
  desiredCapacity: 2,
  machineImage: new BottleRocketImage(stack),
});

app.synth();
