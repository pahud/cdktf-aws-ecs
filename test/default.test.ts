import 'cdktf/lib/testing/adapters/jest';
import { DataAwsAvailabilityZones, AutoscalingGroup, EcsCapacityProvider } from '@cdktf/provider-aws';
import { TerraformStack, Testing } from 'cdktf';
import { Cluster } from '../src';

Testing.setupJest();

describe('Unit testing using snapshots', () => {
  it('should match the snapshot', () => {
    const app = Testing.app();
    const stack = new TerraformStack(app, 'test');
    expect(
      Testing.synthScope(() => {
        new Cluster(stack, 'Cluster');
      })).toMatchSnapshot();
  });
  it('should plan successfully', () => {
    const app = Testing.app();
    const stack = new TerraformStack(app, 'test');
    expect(Testing.fullSynth(stack)).toPlanSuccessfully();
  });
  it('should have a cluster', () => {
    expect(Testing.synthScope((scope) => {
      new Cluster(scope, 'Cluster')
        .addAsgCapacity('Capacity', {
          desiredCapacity: 2,
        });
    })).toHaveDataSource(DataAwsAvailabilityZones);
  });
  it('should have an ASG', () => {
    expect(Testing.synthScope((scope) => {
      new Cluster(scope, 'Cluster')
        .addAsgCapacity('Capacity', {
          desiredCapacity: 2,
        });
    })).toHaveResourceWithProperties(AutoscalingGroup, {
      desired_capacity: 2,
      max_size: 2,
      min_size: 0,
    });
  });
  it('should have a capacity provider', () => {
    expect(Testing.synthScope((scope) => {
      new Cluster(scope, 'Cluster')
        .addAsgCapacity('Capacity', {
          desiredCapacity: 2,
        });
    })).toHaveResource(EcsCapacityProvider);
  });
});

