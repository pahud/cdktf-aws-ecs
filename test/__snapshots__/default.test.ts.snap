// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit testing using snapshots should match the snapshot 1`] = `"{}"`;
